Content/Items/Moonstone/Weapons.CrescentQuarterstaff.cs

```diff
@@ -49,507 +58,661 @@ public override void AddRecipes()
+	{
+		enum AttackType : int
+		{
+			Stab,
+			Spin,
+			UpperCut,
+			Slam
+		}
 
-		private int attackDuration;
+		private const int MAXCHARGE = 10;
 
-		private int currentAttackDuration;
+		private AttackType CurrentAttack
+		{
+			get => (AttackType)Projectile.ai[0];
+			set => Projectile.ai[0] = (float)value;
+		}
 
-		private Vector2 direction;
+		private List<Vector2> cache;
+		private Trail trail;
 
-		private float rotDifference;
+		private ref float charge => ref Projectile.ai[1];
+		private int freezeTimer = 0;
+		private bool active = true;
+		private bool curAttackDone = false;
 
+		private float length = 100;
+		private float initialRotation = 0;
 		private float zRotation = 0;
 
-		private bool FacingRight = false;
-
-		private bool clicked = false;
-
-		private int charge;
+		private int timer = 0;
+		private bool slamCharged = false;
+		private bool slammed = false;
 
-		private float Charge => charge / (float)MAXCHARGE;
-		private bool MouseFacingLeft => Main.MouseWorld.X < Player.Center.X;
-		private bool FirstTickOfSwing
-		{
-			get => Projectile.ai[0] == 0;
-			set => Projectile.ai[0] = value ? 0 : 1;
-		}
+		private Player Player => Main.player[Projectile.owner];
+		private float ArmRotation => Projectile.rotation - ((Player.direction > 0) ? MathHelper.Pi / 3 : MathHelper.Pi * 2 / 3);
+		private float Charge => charge / MAXCHARGE;
+		private Vector2 StaffEnd => Player.GetFrontHandPosition(Player.CompositeArmStretchAmount.Full, ArmRotation) + Vector2.UnitX.RotatedBy(Projectile.rotation) * length * Projectile.scale;
+		private float MeleeSpeed => Player.GetTotalAttackSpeed(DamageClass.Melee);
 
-		Player Player => Main.player[Projectile.owner];
+		private Func<float, float> StabEase = Helper.CubicBezier(0.09f, 0.71f, 0.08f, 1.62f);
+		private Func<float, float> SpinEase = Helper.CubicBezier(0.6f, -0.3f, .3f, 1f);
+		private Func<float, float> UppercutEase = Helper.CubicBezier(0.6f, -0.3f, 0.5f, 0.8f);
+		private Func<float, float> SlamEase = Helper.CubicBezier(0.5f, -1.6f, 0.9f, -1.6f);
 
 		public override string Texture => AssetDirectory.MoonstoneItem + Name;
 
 		public override void SetStaticDefaults()
 		{
 			DisplayName.SetDefault("Crescent Quarterstaff");
-			Main.projFrames[Projectile.type] = 1;
 		}
 
 		public override void SetDefaults()
 		{
 			Projectile.friendly = true;
 			Projectile.DamageType = DamageClass.Melee;
 			Projectile.tileCollide = false;
-			Projectile.Size = new Vector2(130, 130);
+			Projectile.Size = new Vector2(150, 150);
 			Projectile.penetrate = -1;
 			Projectile.usesLocalNPCImmunity = true;
-			Projectile.localNPCHitCooldown = 16;
+			Projectile.localNPCHitCooldown = -1;
 			Projectile.ownerHitCheck = true;
 		}
 
+		public override void OnSpawn(IEntitySource source)
+		{
+			initialRotation = (Main.MouseWorld - Player.MountedCenter).ToRotation();
+			Projectile.scale = Player.GetAdjustedItemScale(Player.HeldItem);
+		}
+
 		public override bool PreAI()
```

Comment:
Do we have much of a reason to use PreAI here over folding this logic into AI or in a more appropriately named method called by AI?